### PR TITLE
Restart rewrite to enable different systems

### PR DESCRIPTION
**Pull request by loganharbour**
refs https://github.com/idaholab/moose/issues/21011

### CODE REVIEW

File: framework/include/base/MooseApp.h
Comment (lindsayad) [inline framework/include/base/MooseApp.h:677]: So a vector of restartable data maps is "the restartable data object"?

File: framework/include/base/MooseApp.h
Comment (lindsayad) [inline framework/include/base/MooseApp.h:717]: and returns the files written?

File: framework/include/base/MooseApp.h
Comment (lindsayad) [inline framework/include/base/MooseApp.h:1128]: this answers my question from earlier

File: framework/include/base/MooseApp.h
Comment (loganharbour) [reply framework/include/base/MooseApp.h:677]: Oh yes, a great example of a bad docstring that has gotten even more bad...

I don't even know if these are used anymore. I've gotten rid of almost all direct data access

**Review by lindsayad — COMMENTED** [review]
Through the headers. Would you prefer that I continue reviewing until 100% done or should I leave time in between reviews?

File: framework/src/restart/RestartableDataWriter.C
Comment (lindsayad) [inline framework/src/restart/RestartableDataWriter.C:16]:
```suggestion
#include "DataIO.h"
#include "RestartableDataMap.h"

#include <fstream>
#include <system_error>
```

File: framework/src/restart/RestartableDataWriter.C
Comment (lindsayad) [inline framework/src/restart/RestartableDataWriter.C:54]: When `dataStore` takes const references we won't need the l-value haha

**Review by lindsayad — COMMENTED** [review]
Changing locations, so publishing current review

**Review by lindsayad — CHANGES_REQUESTED** [review]
I've reviewed all the framework code and I think it looks really great. I think what is needed most at this point is documentation (probably in `restart_recover.md`). What API (`getRestartableData` presumably) should a user call to make sure they can retrievable restartable data that's not declared in the previous simulation?

File: framework/include/problems/FEProblemBase.h
Comment (loganharbour) [reply framework/include/problems/FEProblemBase.h:2120]: I know... this one is kind of important. It's a "managed" value because it needs to get destructed at the proper time, and there's other protected vars that will scream and die otherwise. Hence, I put it as far up as I could.

File: framework/include/problems/FEProblemBase.h
Comment (loganharbour) [reply framework/include/problems/FEProblemBase.h:2201]: these have been removed, not needed anymore

File: framework/include/restart/InputStream.h
Comment (loganharbour) [reply framework/include/restart/InputStream.h:47]: This is useful for error handling. When you pass around one of these streams, you don't know if it's a file or in memory. When you want to throw, if it is a file you can provide a little more context

File: framework/include/restart/RestartableDataIO.h
Comment (loganharbour) [reply framework/include/restart/RestartableDataIO.h:79]: It means that you can't do

```
const std::variant<A, B> data = A{};
data = B{}; // not allowed
```

File: framework/include/restart/RestartableDataMap.h
Comment (loganharbour) [reply framework/include/restart/RestartableDataMap.h:94]: I have no preference here. What would you prefer? I wouldn't object to a standard for what to do with maps with key strings

File: framework/include/restart/RestartableDataMap.h
Comment (loganharbour) [reply framework/include/restart/RestartableDataMap.h:22]: It's insertion ordered. This is so that we can load data in the order that it is instantiated.

I'll add clarity in the docstring

File: framework/src/restart/RestartableDataIO.C
Comment (loganharbour) [reply framework/src/restart/RestartableDataIO.C:14]: It used to be 2, now it's 3. 

Realistically, this should change whenever we make any reasonable changes to data load/store or any of the IO

File: framework/include/restart/RestartableEquationSystems.h
Comment (loganharbour) [reply framework/include/restart/RestartableEquationSystems.h]: Added projections and used it in `System::add_vector` for vectors that we don't have that are added

**Comment by loganharbour** [comment]
Good to go for more.

>  I think what is needed most at this point is documentation (probably in restart_recover.md). What API (getRestartableData presumably) should a user call to make sure they can retrievable restartable data that's not declared in the previous simulation?

I agree. I'll work on writing up some documentation. I'm working with this concurrently on a metadata PR, where we pull in mesh metadata late. There isn't much of a public API at all for this yet as I was trying to iron it out in the follow-on PR. But I can describe most of it at the very least.

The good news is that the late loading is actually tested:
https://github.com/loganharbour/moose/blob/restart_redo_21011/unit/src/RestartableDataIOTest.C

**Comment by lindsayad** [comment]
> Good to go for more.

I don't have more with respect to the code 😄

**Review by lindsayad — DISMISSED** [review]

**Comment by GiudGiud** [comment]
dont merge this unless someone reviews modules

File: test/tests/meshgenerators/file_mesh_generator/tests
Comment (lindsayad) [inline test/tests/meshgenerators/file_mesh_generator/tests:110]:
```suggestion
    requirement = 'The system shall support outputting checkpoints using a consistent directory naming scheme.'
```

File: test/tests/meshgenerators/file_mesh_generator/tests
Comment (lindsayad) [inline test/tests/meshgenerators/file_mesh_generator/tests:121]:
```suggestion
    requirement = 'The system shall support loading checkpointed meshes using the deprecated suffix when a checkpoint exists with the standard suffix.'
```

File: test/tests/restart/restart_transient_from_eigen/tests
Comment (lindsayad) [inline test/tests/restart/restart_transient_from_eigen/tests:9]:
```suggestion
    requirement = 'The system shall support storing the solution of an eigenvalue solve such that it can be restarted later into a non-eigenvalue solve.'
```

File: test/tests/restart/restart_transient_from_eigen/tests
Comment (lindsayad) [inline test/tests/restart/restart_transient_from_eigen/tests:18]:
```suggestion
  []
```

File: unit/src/RestartableDataIOTest.C
Comment (lindsayad) [inline unit/src/RestartableDataIOTest.C:54]:
```suggestion
  auto header_stream = std::make_unique<std::stringstream>();
  auto data_stream = std::make_unique<std::stringstream>();
```

File: unit/src/StringInputStreamTest.C
Comment (lindsayad) [inline unit/src/StringInputStreamTest.C:18]:
```suggestion
  auto input = std::make_unique<std::stringstream>();
```

File: unit/src/StringInputStreamTest.C
Comment (lindsayad) [inline unit/src/StringInputStreamTest.C:30]:
```suggestion
  auto input = std::make_unique<std::stringstream>();
```

File: unit/src/StringInputStreamTest.C
Comment (lindsayad) [inline unit/src/StringInputStreamTest.C:52]:
```suggestion
  auto input = std::make_unique<std::stringstream>();
```

**Review by lindsayad — COMMENTED** [review]
Ok I looked at all the other files

**Review by lindsayad — DISMISSED** [review]

**Review by lindsayad — DISMISSED** [review]

**Comment by lindsayad** [comment]
Oh no, conflicts!

**Comment by YaqiWang** [comment]
Is this the final push? Have the tests I added been included?

**Review by lindsayad — DISMISSED** [review]

**Review by lindsayad — APPROVED** [review]

**Pull request merged**
